smolyay/basis.py

```diff
@@ -403,29 +477,331 @@ def _derivative(self, x):
         float
             Value of the derivative of Trigonometric polynomial.
 
+        n : int, optional
+            order of derivative. Default is 1.
+
         Raises
         ------
         ValueError
             If input is outside the domain `[0, 2\pi]`
         """
         x = numpy.asarray(x)
-        return self.frequency * 1j * numpy.exp(x * self.frequency * 1j)
+        return numpy.exp(x * self.frequency * 1j) * (self.frequency * 1j) ** n
 
 
-class BasisFunctionSet:
+class BasisFunctionSet(collections.abc.Sequence):
     """Set of basis functions and sample points.
 
     Parameters
     ----------
-    basis_functions : list
-        Basis functions in set.
+    basis_functions : list of BasisFunction
+        the basis functions
 
+    Raises
+    ------
+    TypeError
+        Basis functions must have the same domain and typing.
     """
 
-    def __init__(self, basis_functions):
+    def __init__(self, basis_functions=None):
+        self._basis_functions = None
+
+        # check validity of basis functions
+        if isinstance(basis_functions, BasisFunction):
+            basis_functions = [basis_functions]
+        if basis_functions is None:
+            basis_functions = []
+        if len(basis_functions) != 0:
+            domain = basis_functions[0].domain
+            if any(
+                not numpy.array_equal(domain, b.domain)
+                or not type(b) is type(basis_functions[0])
+                for b in basis_functions[1:]
+            ):
+                raise TypeError(
+                    "Basis functions must be be BasisFunction objects with the same domain and typing."
+                )
         self._basis_functions = basis_functions
 
     @property
     def basis_functions(self):
         """list: Basis functions."""
         return self._basis_functions
+
+    @property
+    def domain(self):
+        """numpy.ndarray: Domain of the `basis_functions`"""
+        if self._basis_functions is None or len(self._basis_functions) == 0:
+            raise IndexError("No basis functions to derive a domain.")
```

Comment:
If it works, an attribute error would be better because it means the attribute doesn't exist.
```suggestion
            raise AttributeError("No basis functions to derive a domain.")
```